lang/ir/src/ir/item_impl/mod.rs

```diff
@@ -0,0 +1,355 @@
+        });
+    }
+}
+
+impl ItemImpl {
+    /// Returns `true` if the Rust implementation block is an ink! implementation
+    /// block.
+    ///
+    /// # Note
+    ///
+    /// This is the case if:
+    ///
+    /// - The ink! implementation block has been annotatated as in:
+    ///
+    /// ```
+    /// # use core::convert::TryFrom;
+    /// # <ink_lang_ir::ItemImpl as TryFrom<syn::ItemImpl>>::try_from(syn::parse_quote! {
+    /// #[ink(impl)]
+    /// impl MyStorage {
+    ///     fn my_function(&self) {
+    ///         /* inherent method implementation */
+    ///         unimplemented!()
+    ///     }
+    /// }
+    /// # }).unwrap();
+    /// ```
+    ///
+    /// - Or if any of the ink! implementation block methods do have ink!
+    ///   specific annotations:
+    ///
+    /// ```
+    /// # use core::convert::TryFrom;
+    /// # <ink_lang_ir::ItemImpl as TryFrom<syn::ItemImpl>>::try_from(syn::parse_quote! {
+    /// impl MyStorage {
+    ///     #[ink(constructor)]
+    ///     pub fn my_constructor() -> Self {
+    ///         /* constructor implementation */
+    ///         unimplemented!()
+    ///     }
+    /// }
+    /// # }).unwrap();
+    /// ```
+    ///
+    /// The same rules apply to ink! trait implementation blocks.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error in case of encountered malformed ink! attributes.
+    pub(super) fn is_ink_impl_block(
+        item_impl: &syn::ItemImpl,
+    ) -> Result<bool, syn::Error> {
+        // Quick check in order to efficiently bail out in case where there are
+        // no ink! attributes:
+        if !ir::contains_ink_attributes(&item_impl.attrs)
+            && item_impl
+                .items
+                .iter()
+                .all(|item| !ir::contains_ink_attributes(item.attrs()))
+        {
+            return Ok(false)
+        }
+        // Check if the implementation block itself has been annotated with
+        // `#[ink(impl)]` and return `true` if this is the case.
+        let (ink_attrs, _) = ir::partition_attributes(item_impl.attrs.clone())?;
+        let impl_block_span = item_impl.span();
+        if !ink_attrs.is_empty() {
+            let normalized =
+                ir::InkAttribute::from_expanded(ink_attrs).map_err(|err| {
+                    err.into_combine(format_err!(impl_block_span, "at this invokation",))
+                })?;
+            if normalized
+                .ensure_first(&ir::AttributeArgKind::Implementation)
+                .is_ok()
+            {
+                return Ok(true)
+            }
+        }
+        // Check if any of the implementation block's methods either resembles
+        // an ink! constructor or an ink! message:
+        'outer: for item in &item_impl.items {
+            match item {
+                syn::ImplItem::Method(method_item) => {
+                    if !ir::contains_ink_attributes(&method_item.attrs) {
+                        continue 'outer
```

Comment:
yes and also for robustness in case someone changes the code in the future with additional loops or whatever. I just find the explicitly labeled continue or break statements to be much more readable in general.

Reply:
Yeah it is useful actually to let your eye skip to the loop, just never seen it used before